src/web/components/Figure.tsx

```diff
@@ -150,11 +144,23 @@ const decidePosition = (role: RoleType | 'richLink') => {
 	}
 };
 
+type Props = {
+	children: React.ReactNode;
+	isMainMedia: boolean;
+	role?: RoleType | 'richLink';
+	id?: string;
+
+	// Used to style figures based on role type. Parameterised as this varies
+	// (e.g.) on page layout.
+	roleStyler?: (role: RoleType | 'richLink') => string;
+};
+
 export const Figure = ({
 	role = 'inline',
 	children,
 	id,
 	isMainMedia,
+	roleStyler = standardRoleStyler,
```

Comment:
I am a little worried by this pattern. Once you allow anyone to overload styles, you can quickly start adding a lot of unnecessary complexity. I was trying to find an example of this overload being using in the PR but cannot find one that doesn't use the default `standardRoleStyler`. I would suggest maybe remove `roleStyler` props and then progressively adding styles as we find examples.

Reply:
It is used for the immersive layouts (which is why I introduced it). It is unfortunate, I agree, as it adds messiness/complexity and will probably only get used in this one place.

In general, I think the design mistake here is to couple the layout aspect of elements (the CSS attached to wrapping `Figure`s with the conversion of elements to their React representations. Unfortunately, it seems hard to separate these out easily - partly because the element => JSX.Element step can result in a null (in which case we wouldn't want to wrap in a figure) but it's impossible to identify these cases. Or rather, we could do by adding a load of code.

Reply:
I'd like to try and improve this though in a separate PR once we know if we need this code/interactives are a bit clearer.

Reply:
I am not sure I follow 100%, but you mentioned that in some cases the JSX returned is `null` (I am assuming you are referring to `children` prop in this case), and that in those cases you do not want to return the `figure` HTML. If that is the case it would make more sense to conditionally determine that in the parent component and not here?

Reply:
if we were to have a new `const` to determine styles that are specific to Immersives, would it not make more sent to define that variable here and just conditionally call it if it is interactive. We could call it `interactiveStyles` or something on those lines. This would remove the need for some dynamic prop to set styles

Reply:
Yes that is the alternative, but how to conditionally determine that for each element without essentially reproducing the element logic. Which would add a lot of code and most of it duplicated for the element rendering itself.

I.e. at the moment ElementsRenderer returns: `JSX.Element`, where this can be a null element but impossible to tell. If not null, typically wrapped in `<Figure...>` with some layout/other CSS.
```

This PR parameterises the `Figure` CSS so that it can be customised for different layouts etc. (Required for elements in immersive interactives).

As alternatives to separate out the CSS, we could do:

1. don't wrap in `<Figure ...>` and let the calling code handle that

This would require us to possibly change ElementRenderer to accept all the elements and return an array - that way it could strip out nulls and we could guarantee returned elements required the `<Figure ..` wrapping. However, this is also brittle as it would be difficult to wrap with a different Figure/props based on element type - which we currently do. We could get around this with multiple return etc - e.g. return the element ID along with the element. This would be a bit more code but definitely possible and potentially the 'right' solution here.

2. Duplicate the ElementRenderer code purely for the special case (Immersive interactives) and then modify as required.

This would duplicate a lot of core code though, which is not great.

--

I'm sure there are other options too.

Reply:
could we not filter out "empty" elements using a similar pattern that we do for Numbered lists in the `enhancer`?

Reply:
in the end, I would strongly vouch against this approach, if you really wanted to overload certain styles because the article is `Immersive` I think it would be better to pass down a boolean prop `isImmersive` and then have a css `const` that would conditionally apple relevant styles depending on the `role` and `isImmersive`. This way we would no longer need to pass down `roleStylerroleStyler` function

Reply:
For what it's worth, I'm not a huge fan of this approach either. Passing in functions that return styles goes against the principle of components keeping their styles local and isolated. This was a key driver for the DCR project and I think it's something that is worth continuing to strive for.

Why is it worth it?
The benefit of a component having style logic isolated is
1. It's much easier to reason about what to expect. Everything is local to the file so there's no  need to jump around
2. Any changes only impact the thing you're looking at, there's no style bleed
3. The props can be cleanly and tightly typed. The type of this function is `any` basically

Stepping back, I wonder if the need for this type of change is suggestive that there needs to be a new component? Sometimes the best approach is to create a second comoponent instead of trying to overload the first. I think it's perfectly valid to have `Figure` and `InteractiveFigure` (or whatever name makes sense based on the differences). What do you think?

Reply:
There was a thread about this exact topic https://chat.google.com/room/AAAAWwBdSMs/QJ-dfrnymR4